doc/designs/quic-design/quic-event-loop.md

```diff
@@ -0,0 +1,160 @@
+# Design Problem: Event Loop
+
+Applications need to be able to react to events on:
+
+* Sockets
+* Connections
+* Streams
+* Timers
+
+Only the OpenSSL library can supply the events for Connections and Streams.
+Events on sockets and timers ultimately come from a system source - the
+application may want to receive the event directly from the system, or from
+OpenSSL.
+
+For the MVP a blocking mode on top of one of the options below is likely to be
+required because that’s the default mode for `s_client`.
+
+## Candidate Solution: Application Selects
+
+In this approach the application "selects" (or whatever alternative function
+the applications wishes to use) on the socket and asks OpenSSL to process
+incoming packets when they arrive and tell it about any connection/stream
+events. The application implements the event loop.  Other calls have to be made
+to deal with each of the event possibilities. For timers, OpenSSL supplies a
+function to return how long until the next timer event so that the application
+knows how long to wait.
+
+Pros:
+
+* The application has full control over the event loop and can implement it in
+whichever way most makes sense to the application
+* Closest solution to the current API for SSL/TLS/DTLS
+* Enables the application to handle interaction with the network layer - which
+could be abstracted from OpenSSL via BIOs etc
+
+Cons:
+
+* Requires a lot of "boiler plate" code to get anything working
+* The application has to handle timers itself (unlike applications written for
+SSL/TLS)
+* Not so easy to write the simplest application that want it to "just work"
+
+![alt_text](images/image2.png "Application Selects Sequence Diagram")
+
+## Candidate Solution: OSSL_select()
+
+OpenSSL listens on sockets, de-multiplexes to connections/streams and passes
```

Comment:
As I understand it, right now client side only is the goal, right? Is server side going to be a goal ever? If so, do you want an API that can be easily adapted to server or do you want to write an entirely new API? If you want to be able to add server with minimal changes in the future, I recommend you consider all the complexity involved in mapping a received UDP datagram to the appropriate connection. I would not ever recommend the app be in charge of that logic. Even with a client side scenario, if you support sharing the local socket (which the protocol allows for) with multiple connections, this complexity can exist.

Reply:
> As I understand it, right now client side only is the goal, right? Is server side going to be a goal ever?

Client side is the goal for MVP. Server side will be needed for the final solution.

>  I would not ever recommend the app be in charge of that logic. Even with a client side scenario, if you support sharing the local socket (which the protocol allows for) with multiple connections, this complexity can exist.

Interestingly we've also received [feedback](https://github.com/openssl/openssl/pull/17185#issuecomment-989782783) insisting the opposite, i.e. that the server application *must* be in charge of this.

In practice I think we are going to have to allow for both, i.e. provide the capability for OpenSSL to handle this, but not make it mandatory and allow applications to do it themselves if they wish.

Reply:
I'd be happy to discuss with you further on the intricacies involved here (either on a call, on the QUIC Slack or [MsQuic Discord](https://discord.gg/YGAtCwTSsc)), but bottom line, the logic for mapping a received packet to the appropriate connection, and doing it efficiently, and dealing with things like RSS are very complicated, and generally involve coordination with the logic that generates the connection IDs themselves. Assuming you're not going to expose the ability to control the CID generation, they cannot effectively do this coordination.

BTW, this kind of issue will be just one of probably a hundred nuanced complexity points you will encounter when building a production-grade, general purpose QUIC stack. We've been working on MsQuic for over 4 years now and still aren't done yet. I'm happy to help with advice here, but building a QUIC stack will be a very long road; unless you're able to leverage existing code (also able to help there).

Reply:
Thanks Nick. I might take you up on that offer at some point. :-)